src/coreclr/gc/gc.cpp

```diff
@@ -28200,7 +28376,14 @@ void gc_heap::relocate_survivors (int condemned_gen_number,
                 {
                     current_heap_segment = heap_segment_next_rw (current_heap_segment);
                     current_brick = brick_of (heap_segment_mem (current_heap_segment));
-                    end_brick = brick_of (heap_segment_allocated (current_heap_segment)-1);
+                    end_brick = brick_of (heap_segment_allocated (current_heap_segment) - 1);
+                    if (heap_segment_mem (current_heap_segment) == heap_segment_allocated (current_heap_segment))
+                    {
+                        // the case of an empty region / segment - make sure
+                        // we don't relocate anything, but go to the next region / segment
+                        assert (current_brick == end_brick);
+                        current_brick++;
+                    }
```

Comment:
this can also be eliminated

Reply:
Why is that? Is there another change in your latest checkin that makes sure we never encounter an empty region here?

Reply:
As it turns out, the issue why I added this code is still there. However, my assert was wrong, so I removed it.

Reply:
hmm, that's interesting. I had thought you only needed this because of the bug I fixed in the last checkin. it certainly should be able to handle empty segments or we would have hit this many years ago. so at the very least I don't think this is a problem for segments. I cannot repro this with regions either. the first brick should be -1 (constructed by the plan phase). how did you repro this? could you show the exact command line, the env vars you set and how you built (what features you included)?

Reply:
My repro is with server GC, using the debug build of the master branch of a couple days ago (of course with USE_REGIONS defined, but without any of mymark list changes). Command line is this (copied from your checkin comments for the regions work):

c:\Debuggers\windbg.exe C:\runtime-1\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\CoreRun.exe GcPerfSim.dll -tc 2 -tagb 16 -tlgb 0.05 -lohar 0 -sohsi 10 -lohsi 0 -pohsi 0 -sohpi 0 -lohpi 0 -pohpi 0 -sohfi 0 -lohfi 0 -pohfi 0 -allocType reference -testKind time

Environment variables are these:

I:\GCPerfSim>set complus
complus_gcConcurrent=0
complus_GCRegionsRange=800000000
COMPLUS_gcServer=1

Local variables in relocate_survivors are these:

0:013> dv
              end_brick = 0x127800
                   args = struct SVR::gc_heap::relocate_args
          start_address = 0x000001bd`6e000020 ""
            end_address = 0x000001bd`6e3ff7a8 ""
          condemned_gen = 0x000001bc`c61de6a8
   current_heap_segment = 0x000001c4`c87ea430
          current_brick = 0x127800
                      i = 0n0
                   this = 0x000001bc`c61de550
   condemned_gen_number = 0n2
first_condemned_address = 0x000001bc`e7c00020 "???"
           stop_gen_idx = 0n0

and brick_table[current_brick] is 0.

I think that empty segments work fine without regions, because heap_segment_mem is normally brick aligned, so subtracting one from heap_segment_allocated will result in a smaller brick if the two are the same. With regions, heap_segment_mem is not brick aligned, so subtracting one results in the same brick, which is what I observe.

Reply:
but my point was the brick that corresponds to heap_segment_mem (not (heap_segment_mem - 1)) should be -1, not 0. so it's puzzling why this is 0 for you. I've been running this for a while and cannot repro (I've tried with and without STRESS_REGIONS).  also end_address looks weird - it should be the same as start_address for empty segs. I'm not sure if it's just not displaying the value correctly. I'll see if I can catch you online later today to debug this.

Reply:
oh actually I did mean to mention this - you don't need this anymore with the change in get_free_region, right?

Reply:
Correct - I verified that the change in get_free_region fixes it.

Reply:
And yes, I should take out this code if it's no longer needed. Will double check...